deepspeed/linear/optimized_linear.py

```diff
@@ -0,0 +1,141 @@
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+# DeepSpeed Team
+
+import torch
+import math
+import torch.nn as nn
+import torch.nn.functional as F
+from deepspeed.accelerator import get_accelerator
+import deepspeed.comm as dist
+
+from .config import LoRAConfig, QuantizationConfig
+from .quantization import QuantizedParameter, QuantizedLinear
+
+
+class OptimizedLinear(nn.Module):
+    """
+    Optimized version of nn.Linear that adds features such as:
+      * LoRA w. base weight sharding
+      * FP [6,8,12] quantization
+
+    Arguments:
+        input_dim: Required: size of each input sample
+        output_dim: Required: size of each output sample
+        bias: Optional: If set to False, the layer will not learn an additive bias. Default: False
+        lora_config: Optional: LoRAConfig defining lora features and base-weight-sharding degree
+        quantization_config: Optional: QuantizationConfig defining quantization features
+        dtype: Optional: parameter dtype, only supports bfloat16 currently
+
+    Returns:
+        Returns a new nn.Module depending on the input config. Either native
+        torch.nn.Linear, QuantizedLinear, or the full-featured DSOptimizedLinear.
+    """
+
+    def __new__(self,
+                input_dim: int,
+                output_dim: int,
+                bias: bool = False,
+                lora_config: LoRAConfig = None,
+                quantization_config: QuantizationConfig = None,
+                dtype=torch.bfloat16):
+
+        if lora_config is None and quantization_config is None:
+            # Everything disabled, fall back to normal nn.Linear
+            self = nn.Linear(input_dim, output_dim, bias=bias)
+        elif lora_config:
+            # lora enabled w. optional quant
+            self = DSOptimizedLinear(input_dim=input_dim,
+                                     output_dim=output_dim,
+                                     bias=bias,
+                                     lora_config=lora_config,
+                                     quantization_config=quantization_config,
+                                     dtype=dtype)
+        elif quantization_config:
```

Comment:
It seems that the intention is for user to specify at most one of `lora_config` or `quantization_config`, but not both. However, the current logic allows users to specify both, and future reordering of the conditionals will change code execution in unexpected ways. I feel that a cleaner design would be to pass exactly one quantization config (or None for baseline behavior), and for the logic to figure out which kind of quantization is needed. Happy to chat more in case I am missing something.

Reply:
On further reading, it seems that:
1. lora_config requires (uses) quantization_config 
2. quantization_config is self-contained. 

Can you please correct/confirm my understanding?

Reply:
The if-condition here is a bit strange, I updated the comment around the `LoRAOptimizedLinear` to hopefully better clarify. In general though `OptimizedLinear` should support all 4 combos of lora_config or quant_config being enabled or disabled (TT,TF,FT,FF). In the case of `LoRAOptimizedLinear` it can support LoRA with or without a quant_config. Hope this helps clarify along with the rest of the updates I've just pushed.